base/iterators.jl

```diff
@@ -1428,4 +1428,81 @@ only(x::NamedTuple) = throw(
     ArgumentError("NamedTuple contains $(length(x)) elements, must contain exactly 1 element")
 )
 
+"""
+    unfold(f, init)
+
+Iterate values generated from an initial state and a function.
+
+`f` takes a single value (the state) and returns `(element, nextstate)` or, if it has finished emitting elements, `nothing`.
+
+`Iterators.map` or a generator like `(√x + 2 for x in 1:10)` are iterable or lazy analogs of `for` loops.
+`unfold` is the iterable or lazy analog of the `while` loop.
+
+The "dual" of `foldr`: `foldr` reduces an iterable to a summary value, `unfold` builds an iterable from an initial value.
+
+# Examples
+
+```jldoctest
+julia> using Base.Iterators: unfold
+
+julia> collect(unfold(x -> x > 0 ? (x, x-1) : nothing, 5))
+5-element Vector{Int64}:
+ 5
+ 4
+ 3
+ 2
+ 1
+
+julia> # Same output as digits(6, base=2):
+
+julia> unfold(6) do n
+           n ≠ 0 ? reverse(divrem(n, 2)) : nothing
+       end |> collect
+3-element Vector{Int64}:
+ 0
+ 1
+ 1
+```
+
+# Extended help
+
+The interface for `f` is very similar to the interface required by `iterate`, but `unfold` is simpler to use because it does not require you to define a type.
+You can use this to your advantage when prototyping or writing one-off iterators.
+You may want to define an iterator type instead if you'd like to define `IteratorSize` or `IteratorEltype`; for readability; or if you want to dispatch on the type of your iterator.
+
+`unfold` is related to a `while` loop because:
+
+```julia
+collect(unfold(f, initialstate))
+```
+
+is roughly the same as:
+
+```julia
+acc = []
+state = initialstate
+while true
+    x = f(state)
+    x ≡ nothing && break
+    element, state = x
+    push!(acc, element)
+end
+```
+
+But the `unfold` version may produce a more strictly typed vector and can be easily modified to return a lazy collection by removing `collect()`.
+
+In Haskell and some other functional programming environments, this function is known as `unfoldr`.
+"""
+unfold(f, init) = Unfold(f, init)
+
+struct Unfold{F, T}
+    f::F
+    init::T
+end
+
+iterate(uf::Unfold, state = uf.init) = uf.f(state)
```

Comment:
The iteration protocol is not normally supposed to have an `init` state. Perhaps this would work to easily provide both versions (with and without an `init` argument):
```suggestion
struct Unfold{F, T}
    f::F
    init::T
    Unfold{F, T}(f::F, init::T) where {F, T} = new{F, T}(f, init)
    Unfold(f::F, init::T) where {F, T} = new{F, T}(f, init)
    Unfold{F}(f::F) where {F} = new{F, Union{}}(f)
    Unfold(f::F) where {F} = new{F, Union{}}(f)
end

iterate(uf::Unfold) = isdefined(uf, :init) ? uf.f(uf.init) : uf.f()
iterate(uf::Unfold, state) = uf.f(state)
```

Reply:
This can also be done by removing the `init` field, and using `rest(unfold(f), init)` if you want an initial state.

Reply:
I'm happy to do this if it's desired, but I'd like to point out that all of the other implementations of `unfold` that I linked above require a "seed" or "initial" argument. I think it will be quite rare that `unfold(f)` is wanted rather than `unfold(f, init)` and I would personally prefer that we only define `unfold(f, init)`.

I think Jeff's suggestion of using `rest` is good, so this is how I would implement it if you are sure you want it:

```julia
struct Unfold{F}
    f::F
end

unfold(f) = Unfold(f)
unfold(f, init) = rest(unfold(f), init)

iterate(uf::Unfold) = uf.f()
iterate(uf::Unfold, state) = uf.f(state)
```

And with this text added to the docstring:

> If `init` is not provided, `f` will be called with no arguments on the first iteration, then as normal.

Reply:
While `iterate` does expect two methods to be defined per-type, I think `unfold` is still a useful prototyping tool even if its `f` always receives an initial value.

I think that because you can move the arguments around to always have a state:

```julia
iterate(it::X) => f((it::X,),)
iterate(it::X, state) => f((it::X, state),)
```

Reply:
@JeffBezanson Could you decide here if you'd like what you described or if you were swayed by my argument that other implementations of `unfold` require an `init` value? Thanks!

Reply:
I think `init` is nice in that `f` only needs one method accepting one argument. So I'm ok with that version.